azurerm/resource_arm_iothub_resource_provider.go

```diff
@@ -0,0 +1,264 @@
+		Resourcegroup:  &rg,
+		Name:           &name,
+		Location:       &location,
+		Subscriptionid: &subscriptionID,
+		Sku:            &skuInfo,
+	}
+
+	if tagsI, ok := d.GetOk("tags"); ok {
+		tags := tagsI.(map[string]interface{})
+		desc.Tags = *expandTags(tags)
+	}
+
+	future, err := iothubClient.CreateOrUpdate(ctx, rg, name, desc, updateIoT)
+	if err != nil {
+		return err
+	}
+
+	err = future.WaitForCompletion(ctx, iothubClient.Client)
+	if err != nil {
+		return fmt.Errorf("Error creating or updating IotHub %q (Resource Group %q): %+v", name, rg, err)
+	}
+
+	desc, err = iothubClient.Get(ctx, rg, name)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(*desc.ID)
+	return resourceArmIotHubRead(d, meta)
+
+}
+
+func expandAzureRmIotHubSku(d *schema.ResourceData) devices.IotHubSkuInfo {
+	skuList := d.Get("sku").([]interface{})
+	skuMap := skuList[0].(map[string]interface{})
+	cap := int64(skuMap["capacity"].(int))
+
+	name := skuMap["name"].(string)
+	tier := skuMap["tier"].(string)
+
+	return devices.IotHubSkuInfo{
+		Name:     devices.IotHubSku(name),
+		Tier:     devices.IotHubSkuTier(tier),
+		Capacity: &cap,
+	}
+
+}
+
+func resourceArmIotHubRead(d *schema.ResourceData, meta interface{}) error {
+	armClient := meta.(*ArmClient)
+	iothubClient := armClient.iothubResourceClient
+	ctx := armClient.StopContext
+
+	id, err := parseAzureResourceID(d.Id())
+
+	if err != nil {
+		return err
+	}
+
+	iothubName := id.Path["IotHubs"]
+	desc, err := iothubClient.Get(ctx, id.ResourceGroup, iothubName)
+
+	if err != nil {
+		return err
+	}
+
+	properties := desc.Properties
+
+	keysResp, err := iothubClient.ListKeys(ctx, id.ResourceGroup, iothubName)
+	keyList := keysResp.Response()
+
+	var keys []map[string]interface{}
+	for _, key := range *keyList.Value {
+		keyMap := make(map[string]interface{})
+		keyMap["key_name"] = *key.KeyName
+		keyMap["primary_key"] = *key.PrimaryKey
+		keyMap["secondary_key"] = *key.SecondaryKey
+		keyMap["permissions"] = string(key.Rights)
+		keys = append(keys, keyMap)
+	}
+
+	if err != nil {
+		return fmt.Errorf("Error flattening `shared_access_policy` in IoTHub %q: %+v", iothubName, err)
+	}
+
+	d.Set("shared_access_policy", keys)
+	d.Set("hostname", properties.HostName)
+	d.Set("type", desc.Type)
+	flattenAndSetTags(d, &desc.Tags)
+
+	return nil
+}
+
+func resourceArmIotHubDelete(d *schema.ResourceData, meta interface{}) error {
+
+	id, err := parseAzureResourceID(d.Id())
+
+	if err != nil {
+		return err
+	}
+
+	armClient := meta.(*ArmClient)
+	iothubClient := armClient.iothubResourceClient
+	ctx := armClient.StopContext
+
+	iotHubName := id.Path["IotHubs"]
+
+	future, err := iothubClient.Delete(ctx, id.ResourceGroup, iotHubName)
+	if err != nil {
+		return err
+	}
+
+	err = future.WaitForCompletion(ctx, iothubClient.Client)
+	if err != nil {
+		if response.WasNotFound(future.Response()) {
+			return nil
+		}
+
+		return fmt.Errorf("Error waiting for the deletion of IoTHub %q", iotHubName)
```

Comment:
There is a timeout issue here. IoTHub is deleted in Azure, but terraform times out with an error destroying resource.

Reply:
so this timeout is configured in the client initialisation above: https://github.com/terraform-providers/terraform-provider-azurerm/pull/887/files/3e927a6d96d0c91b3dc132d9b9e83f4b1fc96650#diff-848235249455ee48aaf3ea5d431960aeR597

we can actually swap that out for this function:

https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/config.go#L435
[which sets a 60m polling timeout](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/config.go#L190), which should solve this issue. There's a separate task to move everything over to that at some point, but any changes to `azure/config.go` cause merge conflicts, so we're trying to do it piecemeal rather than bit by bit

Hope that helps :)

Reply:
This is very helpful, thank you!